planning/autoware_path_generator/src/utils.cpp

```diff
@@ -258,47 +237,308 @@ std::vector<std::pair<lanelet::ConstPoints3d, std::pair<double, double>>> get_wa
+          return s_start_edge;
+        }
+        lanelet::BasicPoints2d start_edge_intersections;
+        boost::geometry::intersection(start_edge, bound, start_edge_intersections);
+        for (const auto & intersection : start_edge_intersections) {
+          if (boost::geometry::equals(intersection, bound.front())) {
+            continue;
+          }
+          const auto s = lanelet::geometry::toArcCoordinates(bound, intersection).length;
+          s_start_edge = s_start_edge ? std::min(*s_start_edge, s) : s;
+        }
+        return s_start_edge;
+      };
+    const auto s_left_bound = [&]() {
+      auto s = get_start_edge_intersection_arc_length(cropped_left_bound);
+      if (s) {
+        *s += s_start_left_bound;
+      }
+      return s;
+    }();
+    const auto s_right_bound = [&]() {
+      auto s = get_start_edge_intersection_arc_length(cropped_right_bound);
+      if (s) {
+        *s += s_start_right_bound;
+      }
+      return s;
+    }();
+    const auto [s_left, s_right] =
+      get_arc_length_on_centerline(lanelet_sequence, s_left_bound, s_right_bound);
+    const auto s_start_edge = [&]() {
+      if (s_left && s_right) {
+        return std::min(s_left, s_right);
+      }
+      return s_left ? s_left : s_right;
+    }();
+    if (s_intersection && s_start_edge) {
+      s_intersection = std::min(s_intersection, s_start_edge);
+    } else if (s_start_edge) {
+      s_intersection = s_start_edge;
+    }
+  }
+
+  // intersection between start edge of drivable area and center line
+  {
+    std::optional<double> s_start_edge = std::nullopt;
+    lanelet::BasicPoints2d start_edge_intersections;
+    boost::geometry::intersection(start_edge, cropped_centerline, start_edge_intersections);
+    for (const auto & intersection : start_edge_intersections) {
+      auto s = lanelet::geometry::toArcCoordinates(cropped_centerline, intersection).length;
+      // Ignore intersections near the beginning of the centerline.
+      // It is impossible to make a turn shorter than the vehicle_length, so use it as a threshold.
+      if (s < vehicle_length) continue;
+      s += s_start;
+      s_start_edge = s_start_edge ? std::min(*s_start_edge, s) : s;
+    }
+    if (s_intersection && s_start_edge) {
+      s_intersection = std::min(s_intersection, s_start_edge);
+    } else if (s_start_edge) {
+      s_intersection = s_start_edge;
+    }
+  }
+
+  return s_intersection;
+}
+
+std::optional<double> get_first_self_intersection_arc_length(
+  const lanelet::BasicLineString2d & line_string)
+{
+  if (line_string.size() < 3) {
+    return std::nullopt;
+  }
+
+  const auto tree = lanelet::geometry::internal::makeIndexedSegmenTree(line_string);
+  std::optional<lanelet::geometry::internal::SelfIntersectionLong> first_self_intersection_long =
+    std::nullopt;
+  double s = 0.;
+
+  for (size_t i = 1; i < line_string.size() - 1; ++i) {
+    if (first_self_intersection_long && i == first_self_intersection_long->idx) {
```

Comment:
```suggestion
    if (first_self_intersection_long && i == first_self_intersection_long->idx + 1) {
```

segment index is same tome line string point index - 1

Reply:
before:

https://github.com/user-attachments/assets/ecbbc9f7-e18b-4d1e-865b-ecfefb2de572

after: 

https://github.com/user-attachments/assets/d7c19257-e580-4cce-bd87-935450c71d3e

ok with another case too



https://github.com/user-attachments/assets/423f711a-f353-44a4-bad9-c50dbb04199e